### PR TITLE
Using the repl policy in container to choose partition

### PR DESCRIPTION
**Pull request by vgkholla**
This commit adds the ability in the Router to choose a partition to write to based on the replication policy defined in the container

### CODE REVIEW

File: ambry-api/src/test/java/com.github.ambry/rest/MockBlobStorageServiceFactory.java [hunk 1]
```diff
@@ -31,7 +31,7 @@
   private final Router router;
 
   public MockBlobStorageServiceFactory(VerifiableProperties verifiableProperties, ClusterMap clusterMap,
-      RestResponseHandler restResponseHandler, Router router, Notifier notifier) {
+      RestResponseHandler restResponseHandler, Router router, AccountService accountService) {
     this.verifiableProperties = verifiableProperties;
```
Comment (jsjtzyy) [inline ambry-api/src/test/java/com.github.ambry/rest/MockBlobStorageServiceFactory.java]: So, the `accountService` is temporarily not used in this PR?   (also for `clusterMap`)

File: ambry-clustermap/src/test/java/com.github.ambry.clustermap/MockClusterMap.java [hunk 2]
```diff
@@ -203,7 +216,11 @@ public byte getLocalDatacenterId() {
 
   @Override
   public String getDatacenterName(byte id) {
-    return null;
+    int idx = (int) id;
+    if (idx < 0 || idx >= dataCentersInClusterMap.size()) {
+      return null;
+    }
+    return dataCentersInClusterMap.get(id);
   }
```
Comment (jsjtzyy) [inline ambry-clustermap/src/test/java/com.github.ambry.clustermap/MockClusterMap.java]: `dataCentersInClusterMap.get(idx)`?

File: ambry-clustermap/src/test/java/com.github.ambry.clustermap/MockClusterMap.java [hunk 3]
```diff
@@ -116,12 +121,20 @@ public MockClusterMap(boolean enableSSLPorts, int numNodes, int numMountPointsPe
...
       }
     }
+    if (numNodes > 3 && dcToDataNodes.get(localDatacenterName).size() > 2) {
+      // create one "special partition" that has 3 replicas in the local datacenter (as configured on startup) and
+      // 2 everywhere else
+      List<MockDataNodeId> nodeIds = new ArrayList<>();
+      dcToDataNodes.forEach(
+          (s, mockDataNodeIds) -> nodeIds.addAll(mockDataNodeIds.subList(0, localDatacenterName.equals(s) ? 3 : 2)));
+      PartitionId id = new MockPartitionId(partitionId, SPECIAL_PARTITION_CLASS, nodeIds, 0);
+      partitions.put(partitionId, id);
```
Comment (jsjtzyy) [inline ambry-clustermap/src/test/java/com.github.ambry.clustermap/MockClusterMap.java]: In the statement `PartitionId id = new MockPartitionId(partitionId, SPECIAL_PARTITION_CLASS, nodeIds, 0);`,  a new `partitionId` is created and mounted on `mountPathIndexToUse = 0`.

File: ambry-frontend/src/main/java/com.github.ambry.frontend/AmbryBlobStorageServiceFactory.java [hunk 4]
```diff
@@ -51,11 +49,11 @@
...
    * @param router the {@link Router} to use.
-   * @param notifier the {@link Notifier} to use.
+   * @param accountService the {@link AccountService} to use.
    * @throws IllegalArgumentException if any of the arguments are null.
    */
   public AmbryBlobStorageServiceFactory(VerifiableProperties verifiableProperties, ClusterMap clusterMap,
-      RestResponseHandler responseHandler, Router router, Notifier notifier) {
+      RestResponseHandler responseHandler, Router router, AccountService accountService) {
     if (verifiableProperties == null || clusterMap == null || responseHandler == null || router == null) {
       throw new IllegalArgumentException("Null arguments were provided during instantiation!");
```
Comment (jsjtzyy) [inline ambry-frontend/src/main/java/com.github.ambry.frontend/AmbryBlobStorageServiceFactory.java]: would be better if we check `accountService == null` because `accountService` is directly passed to `getBlobStorageService` method.

File: ambry-frontend/src/test/java/com.github.ambry.frontend/AmbryBlobStorageServiceFactoryTest.java [hunk 5]
```diff
@@ -94,7 +97,7 @@ public void getAmbryBlobStorageServiceFactoryWithBadInputTest() throws Exception
     // Router null.
     try {
```
Comment (jsjtzyy) [inline ambry-frontend/src/test/java/com.github.ambry.frontend/AmbryBlobStorageServiceFactoryTest.java]: also test `accountService` null?

File: ambry-replication/src/test/java/com.github.ambry.replication/ReplicationTest.java [hunk 6]
```diff
@@ -307,8 +307,11 @@ public void replicationPauseTest() throws Exception {
   @Test
   public void replicaThreadTest() throws Exception {
     MockClusterMap clusterMap = new MockClusterMap();
-    Host localHost = new Host(clusterMap.getDataNodeIds().get(0), clusterMap);
-    Host remoteHost = new Host(clusterMap.getDataNodeIds().get(1), clusterMap);
+    // to make sure we select hosts with the SPECIAL_PARTITION_CLASS, pick hosts from the replicas of that partition
+    PartitionId specialPartitionId = clusterMap.getWritablePartitionIds(MockClusterMap.SPECIAL_PARTITION_CLASS).get(0);
+    // these hosts have replicas of the "special" partition and all the other partitions.
```
Comment (jsjtzyy) [inline ambry-replication/src/test/java/com.github.ambry.replication/ReplicationTest.java]: I am thinking if we need to check `specialPartitionId == 0`? Considering the `getWritablePartitionIds` depends on `partitionSelectionHelper`, and the latter one relies on `partitions.values()`. Hence, we cannot guarantee `partitions` contains the "special partition".

File: ambry-router/src/test/java/com.github.ambry.router/PutManagerTest.java [hunk 7]
```diff
@@ -621,6 +628,58 @@ public void testChunkFillerSleep() throws Exception {
...
+    // composite
+    sizeToChunkCount.put(chunkSize * count, count + 1);
+    for (Map.Entry<Integer, Integer> sizeAndChunkCount : sizeToChunkCount.entrySet()) {
+      for (Map.Entry<Container, String> containerAndPartClass : containerToPartClass.entrySet()) {
+        requestAndResultsList.clear();
+        requestAndResultsList.add(new RequestAndResult(sizeAndChunkCount.getKey(), containerAndPartClass.getKey()));
+        // since the puts are processed one at a time, it is fair to check the last partition class set
+        mockClusterMap.clearLastNRequestedPartitionClasses();
+        submitPutsAndAssertSuccess(true);
+        checkLastRequestPartitionClasses(sizeAndChunkCount.getValue(), containerAndPartClass.getValue());
```
Comment (jsjtzyy) [inline ambry-router/src/test/java/com.github.ambry.router/PutManagerTest.java]: Prefer to move the comment `// since the puts are processed one at a time, it is fair to check the last partition class set` to the position before `checkLastRequestPartitionClasses`

File: ambry-router/src/test/java/com.github.ambry.router/PutManagerTest.java [hunk 8]
```diff
@@ -621,6 +628,58 @@ public void testChunkFillerSleep() throws Exception {
...
+      }
+    }
+
+    // exception if there is no partition class that conforms to the replication policy
+    String nonExistentClass = UtilsTest.getRandomString(3);
+    accountService.addReplicationPolicyToContainer(container, nonExistentClass);
+    requestAndResultsList.clear();
+    requestAndResultsList.add(new RequestAndResult(chunkSize, container));
```
Comment (jsjtzyy) [inline ambry-router/src/test/java/com.github.ambry.router/PutManagerTest.java]: Do we also need to test `nonExistentClass` with composite blob? (looks like only simple blob is tested here)

File: ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerTestFramework.java [hunk 9]
```diff
@@ -192,8 +197,19 @@ static Properties getRouterProperties(String routerDatacenter) {
    * @param blobId the blobId
    * @param operationName a name for the operation being checked
    */
```
Comment (jsjtzyy) [inline ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerTestFramework.java]: `@param properties`?

File: ambry-api/src/test/java/com.github.ambry/account/InMemAccountService.java [hunk 10]
```diff
@@ -149,6 +148,20 @@ public synchronized Account createAndAddRandomAccount() {
...
+   * @param replicationPolicy the replication policy to add to {@code container}.
+   * @return the new container object with the appropriate replication policy if the operation suceeded. {@code null}
+   * otherwise
+   */
+  public Container addReplicationPolicyToContainer(Container container, String replicationPolicy) {
+    Container newContainer = new ContainerBuilder(container).setReplicationPolicy(replicationPolicy).build();
+    Account account = getAccountById(container.getParentAccountId());
+    Account newAccount = new AccountBuilder(account).addOrUpdateContainer(newContainer).build();
+    return updateAccounts(Collections.singleton(newAccount)) ? newContainer : null;
+  }
```
[the remark this answers is not in the document]
Comment (vgkholla) [reply ambry-api/src/test/java/com.github.ambry/account/InMemAccountService.java]: Right - in this implementation, it does. This piece of code is not making the assumption in case the impl of `updateAccounts()` changes in the future.

File: ambry-api/src/test/java/com.github.ambry/rest/MockBlobStorageServiceFactory.java (hunk 1 above)
Comment (vgkholla) [reply ambry-api/src/test/java/com.github.ambry/rest/MockBlobStorageServiceFactory.java]: Yeah - none of the tests needed it. If necessary, can be used in the future

File: ambry-clustermap/src/test/java/com.github.ambry.clustermap/MockClusterMap.java (hunk 2 above)
Comment (vgkholla) [reply ambry-clustermap/src/test/java/com.github.ambry.clustermap/MockClusterMap.java]: Changed

File: ambry-clustermap/src/test/java/com.github.ambry.clustermap/MockClusterMap.java (hunk 3 above)
Comment (vgkholla) [reply ambry-clustermap/src/test/java/com.github.ambry.clustermap/MockClusterMap.java]: It does - It doesn't look like any of the tests care but I'll take a look at trying to maintain that invariant.

On the flip side, you can say that the callers want `numStoresPerMountPoint` "default" stores per mount point. Let me think over this.

File: ambry-frontend/src/main/java/com.github.ambry.frontend/AmbryBlobStorageServiceFactory.java (hunk 4 above)
Comment (vgkholla) [reply ambry-frontend/src/main/java/com.github.ambry.frontend/AmbryBlobStorageServiceFactory.java]: Also added a test

File: ambry-frontend/src/test/java/com.github.ambry.frontend/AmbryBlobStorageServiceFactoryTest.java (hunk 5 above)
Comment (vgkholla) [reply ambry-frontend/src/test/java/com.github.ambry.frontend/AmbryBlobStorageServiceFactoryTest.java]: Done

File: ambry-replication/src/test/java/com.github.ambry.replication/ReplicationTest.java (hunk 6 above)
Comment (vgkholla) [reply ambry-replication/src/test/java/com.github.ambry.replication/ReplicationTest.java]: That's true - however, `PartitionSelectionHelper::getPartitionsInClass()` throws an `IllegalArgumentException` if there are no partitions for the class provided - so that should take care of that. Even if that implementation were to change (to say return a null/empty list), since we do a `get(0)` here, there will still be an exception. That should be good enough - what do you think?

File: ambry-rest/src/main/java/com.github.ambry.rest/RestServer.java [hunk 11]
```diff
@@ -288,10 +299,16 @@ public void shutdown() {
       long routerCloseTime = System.currentTimeMillis();
       elapsedTime = routerCloseTime - responseHandlerShutdownTime;
       logger.info("Router close took {} ms", elapsedTime);
-      restServerMetrics.routerCloseTime.update(elapsedTime);
+      restServerMetrics.routerCloseTimeInMs.update(elapsedTime);
+
+      accountService.close();
+      long accountServiceCloseTime = System.currentTimeMillis();
+      elapsedTime = accountServiceCloseTime - responseHandlerShutdownTime;
```
[the remark this answers is not in the document]
Comment (vgkholla) [reply ambry-rest/src/main/java/com.github.ambry.rest/RestServer.java]: Good catch - fixed

File: ambry-clustermap/src/test/java/com.github.ambry.clustermap/MockClusterMap.java (hunk 3 above)
Comment (vgkholla) [reply ambry-clustermap/src/test/java/com.github.ambry.clustermap/MockClusterMap.java]: Done - changed docs to reflect all the changes

File: ambry-router/src/test/java/com.github.ambry.router/PutManagerTest.java (hunk 7 above)
Comment (vgkholla) [reply ambry-router/src/test/java/com.github.ambry.router/PutManagerTest.java]: Done

File: ambry-router/src/test/java/com.github.ambry.router/PutManagerTest.java (hunk 8 above)
Comment (vgkholla) [reply ambry-router/src/test/java/com.github.ambry.router/PutManagerTest.java]: I don't think so - the code path is the same

File: ambry-server/src/integration-test/java/com.github.ambry.server/MockCluster.java [hunk 12]
```diff
@@ -51,43 +52,43 @@
...
+    List<MockDataNodeId> dataNodes = clusterMap.getDataNodes();
+    for (int i = 0; i < dataNodes.size(); i++) {
+      if (sslEnabledDataCenterList != null) {
+        dataNodes.get(i).setSslEnabledDataCenters(sslEnabledDataCenterList);
       }
-    } catch (InstantiationException e) {
-      // clean up other servers which was started already
-      cleanup();
-      throw e;
+      initializeServer(dataNodes.get(i), sslProps, enableHardDeletes, prefetchDataNodeIndex == i ? true : false,
```
[the remark this answers is not in the document]
Comment (vgkholla) [reply ambry-server/src/integration-test/java/com.github.ambry.server/MockCluster.java]: Done

File: ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerTestFramework.java (hunk 9 above)
Comment (vgkholla) [reply ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerTestFramework.java]: Added

**Comment by vgkholla** [comment]
@jsjtzyy Comments addressed.

File: ambry-clustermap/src/test/java/com.github.ambry.clustermap/MockClusterMap.java [hunk 13]
```diff
@@ -50,78 +50,98 @@
...
+  private final MockPartitionId specialPartition;
+
   /**
-   * The default constructor sets up a 9 node cluster with 3 mount points in each, with 3 partitions/replicas per
-   * mount point which amount to a total of 9 replicas per node and 81 replicas across the cluster. If this cluster map
-   * is going to be used to start a cluster, use it judiciously to avoid resource consumption issues on the test
-   * machine.
+   * The default constructor sets up a 9 node cluster with 3 mount points in each, with 3 default partitions/replicas
+   * per mount point. It will also add replicas for a "special" partition. The distribution of these replicas will be
+   * 3 in the chosen "local" datacenter and 2 everywhere else. This will amount to a total of 9/10 replicas per node and
```
Comment (jsjtzyy) [inline ambry-clustermap/src/test/java/com.github.ambry.clustermap/MockClusterMap.java]: Minor: maybe change `9/10 replicas` to `9 or 10 replicas`? (to avoid ambiguity)

**Comment by vgkholla** [comment]
@cgtz Reminder to review

File: ambry-router/src/main/java/com.github.ambry.router/PutManager.java [hunk 14]
```diff
@@ -151,14 +157,31 @@ public void onEvent(ByteBufferAsyncWritableChannel.EventType e) {
...
     putOperations.add(putOperation);
     putOperation.startReadingFromChannel();
   }
 
+  /**
+   * @param blobProperties the properties of the blob being put
+   * @return the partition class as required by the properties
+   */
+  private String getPartitionClass(BlobProperties blobProperties) {
+    String partitionClass = defaultPartitionClass;
```
Comment (cgtz) [inline ambry-router/src/main/java/com.github.ambry.router/PutManager.java]: When implementing encryption support, Siva chose to not pass AccountService into the router and keep `isEncrypted` in the `BlobProperties`. Why did we choose to handle replication policy differently this time? Is it because we do not intend to persist the replication policy in the on-disk blob properties?

File: ambry-router/src/test/java/com.github.ambry.router/PutManagerTest.java [hunk 15]
```diff
@@ -701,7 +760,7 @@ private void setupEncryptionCast(VerifiableProperties vProps) throws GeneralSecu
    * @param shouldCloseRouterAfter whether the router should be closed after the operation.
    */
   private void submitPutsAndAssertSuccess(boolean shouldCloseRouterAfter) throws Exception {
-    submitPut().await();
+    submitPut().await(5, TimeUnit.SECONDS);
```
Comment (cgtz) [inline ambry-router/src/test/java/com.github.ambry.router/PutManagerTest.java]: Create a constant for the timeout length?

File: ambry-server/src/integration-test/java/com.github.ambry.server/MockCluster.java [hunk 16]
```diff
@@ -280,6 +287,14 @@ public void onBlobDeleted(String blobId, String serviceId) {
   @Override
   public synchronized void onBlobReplicaCreated(String sourceHost, int port, String blobId,
       BlobReplicaSourceType sourceType) {
+    int numberOfReplicas;
+    try {
+      BlobId blobIdObj = new BlobId(blobId, clusterMap);
+      PartitionId partitionId = blobIdObj.getPartition();
+      numberOfReplicas = partitionId.getReplicaIds().size();
+    } catch (Exception e) {
+      throw new IllegalArgumentException("Invalid blob ID: " + blobId);
```
Comment (cgtz) [inline ambry-server/src/integration-test/java/com.github.ambry.server/MockCluster.java]: pass in the cause exception too.

File: ambry-router/src/main/java/com.github.ambry.router/PutManager.java (hunk 14 above)
Comment (vgkholla) [reply ambry-router/src/main/java/com.github.ambry.router/PutManager.java]: I think with encrypted, `isEncrypted` _is_ a part of the properties (since the container may switch it on/off but the blob remains encrypted/unencrypted) so it may not be apples to apples. 

In this case, the replication policy is actually a property of the container and not the blob i.e. if the replication policy were to change, the partition that the blob is on will now conform to the new replication policy (and hence the blob).

Comment (cgtz) [reply ambry-router/src/main/java/com.github.ambry.router/PutManager.java]: The replication policy is mostly a property of the partition num chosen, right? If my understanding is correct, changing the container's replicationPolicy would only affect new blobs uploaded to the container, like with encryption. However, I understand that you could potentially change the partition class of the partition a specific blob was uploaded to, so its also not really a "blob" level property like encryption

Comment (vgkholla) [reply ambry-router/src/main/java/com.github.ambry.router/PutManager.java]: Correct - not a blob level property

File: ambry-router/src/test/java/com.github.ambry.router/PutManagerTest.java (hunk 15 above)
Comment (vgkholla) [reply ambry-router/src/test/java/com.github.ambry.router/PutManagerTest.java]: Done

File: ambry-server/src/integration-test/java/com.github.ambry.server/MockCluster.java (hunk 16 above)
Comment (vgkholla) [reply ambry-server/src/integration-test/java/com.github.ambry.server/MockCluster.java]: Done

**Comment by vgkholla** [comment]
@cgtz Comments addressed
